src/ds/d400/d400-options.h

```diff
@@ -100,52 +109,101 @@ namespace librealsense
+            if( _ae_gain_limits_new_opcode )
             {
-                cmd.param2 = static_cast<int>(set_limit);
-                cmd.param3 = *(reinterpret_cast<uint32_t*>(ret.data() + 4));
+                // get structure: min ae, min gain, max ae, max gain
+                command cmd_get( ds::GETAELIMITS );
+                std::vector< uint8_t > ret = _hwm.send( cmd_get );
+                if( ret.empty() )
+                    throw invalid_value_exception( "auto_exposure_limit_option::query result is empty!" );
+
+                // set structure: min ae, max ae, min gain, max gain
+                command cmd( ds::SETAELIMITS );
+                cmd.param1 = 0;
+                cmd.param2 = *( reinterpret_cast< uint32_t * >( ret.data() + 8 ) );
+                cmd.param3 = 0;
+                cmd.param4 = static_cast< int >( set_limit );
+                if( _option == RS2_OPTION_AUTO_EXPOSURE_LIMIT_TOGGLE )
+                {
+                    cmd.param2 = static_cast< int >( set_limit );
+                    cmd.param4 = *( reinterpret_cast< uint32_t * >( ret.data() + 12 ) );
+                }
+                _hwm.send( cmd );
             }
-            _hwm.send(cmd);
+            else
+            {
+                command cmd_get( ds::AUTO_CALIB );
+                cmd_get.param1 = 5;
+                std::vector< uint8_t > ret = _hwm.send( cmd_get );
+                if( ret.empty() )
+                    throw invalid_value_exception( "auto_exposure_limit_option::query result is empty!" );
+
+                command cmd( ds::AUTO_CALIB );
+                cmd.param1 = 4;
+                cmd.param2 = *( reinterpret_cast< uint32_t * >( ret.data() ) );
+                cmd.param3 = static_cast< int >( set_limit );
+                if( _option == RS2_OPTION_AUTO_EXPOSURE_LIMIT_TOGGLE )
+                {
+                    cmd.param2 = static_cast< int >( set_limit );
+                    cmd.param3 = *( reinterpret_cast< uint32_t * >( ret.data() + 4 ) );
+                }
+                _hwm.send( cmd );
+            }
+            
         };
         virtual float query() const override 
-        { 
-            auto offset = 0;
-            if (_option == RS2_OPTION_AUTO_GAIN_LIMIT_TOGGLE)
-                offset = 4;
-            command cmd(ds::AUTO_CALIB);
-            cmd.param1 = 5;
-            auto res = _hwm.send(cmd);
-            if (res.empty())
-                throw invalid_value_exception("auto_exposure_limit_option::query result is empty!");
-
-            auto limit_val = static_cast<float>(*(reinterpret_cast<uint32_t*>(res.data() + offset)));
-            if (limit_val == 0)
+        {
+            float limit_val;
+            int offset = 0;
+            std::vector< uint8_t > res;
+            if( _ae_gain_limits_new_opcode )
+            {
+               offset = 8;
+                if( _option == RS2_OPTION_AUTO_GAIN_LIMIT_TOGGLE )
```

Comment:
It might be better to have `ae_limits_option` and `gain_limits_option` inheriting `limits_option`, each calling a shared function passing a different offset value.